pkg/volume/util/goroutinemap/goroutinemap.go

```diff
@@ -0,0 +1,252 @@
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName,
+	operationFunc func() error) error {
+	grm.lock.Lock()
+	defer grm.lock.Unlock()
+
+	var previousOp operation
+	opExists := false
+	previousOpIndex := -1
+	for i, existingOp := range grm.operations {
+		if existingOp.volumeName != volumeName {
+			// No match, keep searching
+			continue
+		}
+
+		if existingOp.podName != emptyUniquePodName &&
+			podName != emptyUniquePodName &&
+			existingOp.podName != podName {
+			// No match, keep searching
+			continue
+		}
+
+		// Match
+		opExists = true
+		previousOp = existingOp
+		previousOpIndex = i
+		break
+	}
+
+	if opExists {
+		// Operation already exists
+		if previousOp.operationPending {
+			// Operation is pending
+			operationName := getOperationName(volumeName, podName)
+			return common.NewAlreadyExistsError(operationName)
+		}
+
+		if !previousOp.expBackoff.SafeToRetry() {
+			operationName := getOperationName(volumeName, podName)
+			return common.NewExponentialBackoffError(
+				operationName, previousOp.expBackoff)
+		}
+	}
+
+	if opExists {
+		// Update existing operation to mark as pending.
+		grm.operations[previousOpIndex].operationPending = true
+		grm.operations[previousOpIndex].volumeName = volumeName
+		grm.operations[previousOpIndex].podName = podName
+	} else {
+		// Create a new operation
+		grm.operations = append(grm.operations,
+			operation{
+				operationPending: true,
+				volumeName:       volumeName,
+				podName:          podName,
+				expBackoff:       previousOp.expBackoff,
+			})
+	}
+
+	go func() (err error) {
+		// Handle unhandled panics (very unlikely)
+		defer k8sRuntime.HandleCrash()
+		// Handle completion of and error, if any, from operationFunc()
+		defer grm.operationComplete(volumeName, podName, &err)
+		// Handle panic, if any, from operationFunc()
+		operationName := getOperationName(volumeName, podName)
+		defer common.RecoverFromPanic(operationName, &err)
+		return operationFunc()
+	}()
+
+	return nil
+}
+
+func (grm *goRoutineMap) getOperation(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName) (uint, error) {
+
+	for i, op := range grm.operations {
+		if op.volumeName == volumeName &&
+			op.podName == podName {
+			return uint(i), nil
+		}
+	}
+
+	logOperationName := getOperationName(volumeName, podName)
+	return 0, fmt.Errorf("Operation %q not found.", logOperationName)
+}
+
+func (grm *goRoutineMap) deleteOperation(
+	volumeName api.UniqueVolumeName,
+	podName types.UniquePodName) {
+
+	opIndex := -1
+	for i, op := range grm.operations {
+		if op.volumeName == volumeName &&
+			op.podName == podName {
+			opIndex = i
+			break
+		}
+	}
+
+	// Delete index without preserving order
+	grm.operations[opIndex] = grm.operations[len(grm.operations)-1]
+	grm.operations = grm.operations[:len(grm.operations)-1]
+}
+
+func (grm *goRoutineMap) operationComplete(
+	volumeName api.UniqueVolumeName, podName types.UniquePodName, err *error) {
+	defer grm.cond.Signal()
```

Comment:
My brain twisted a bit at this - just to verify, we're going to take and hold lock, but cond will be signaled inside the lock?  So lock dominates cond here as well as in wait?